lib/auth/grpcserver.go

```diff
@@ -1065,6 +1066,271 @@ func (g *GRPCServer) DeleteAllKubeServices(ctx context.Context, req *proto.Delet
 	return &empty.Empty{}, nil
 }
 
+func (g *GRPCServer) AddMFADevice(stream proto.AuthService_AddMFADeviceServer) error {
+	ctx := stream.Context()
+	actx, err := g.authenticate(ctx)
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	auth := actx.authServer
+	user := actx.User.GetName()
+
+	// The RPC is streaming both ways and the message sequence is:
+	// (-> means client-to-server, <- means server-to-client)
+	//
+	// 1. -> Init
+	// 2. <- ExistingMFAChallenge
+	// 3. -> ExistingMFAResponse
+	// 4. <- NewMFARegisterChallenge
+	// 5. -> NewMFARegisterResponse
+	// 6. <- Ack
+
+	// 1. receive client Init
+	req, err := stream.Recv()
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	initReq := req.GetInit()
+	if initReq == nil {
+		return trail.ToGRPC(trace.BadParameter("expected AddMFADeviceRequestInit, got %T", req))
+	}
+	devs, err := auth.GetMFADevices(ctx, user)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, d := range devs {
+		if d.Metadata.Name == initReq.DeviceName {
+			return trail.ToGRPC(trace.AlreadyExists("MFA device named %q already exists", d.Metadata.Name))
+		}
+	}
+
+	// 2. send MFAAuthenticateChallenge
+	// Note: authChallenge may be empty if this user has no existing MFA devices.
+	authChallenge, err := auth.mfaAuthChallenge(ctx, user)
+	if err != nil {
+		return trail.ToGRPC(err)
+	}
+	if err := stream.Send(&proto.AddMFADeviceResponse{
```

Comment:
If there are no challenges in `authChallenge`, should this roundtrip `<- ExistingMFAChallenge, -> ExistingMFAResponse` still be done or can it be short-circuited?

Reply:
I was debating this too.
On the one hand - yes, it eliminates a roundtrip.
On the other hand - this would make the client-side code more complicated and create possibility for bugs.

I opted to keep the protocol in this RPC straightforward, without any conditional messages.

But I'm open to changing this, if you think the extra roundtrip is too big of a price to pay.

Reply:
Nah, I think it's fine. I was mostly curious how the client would handle scenario with empty challenges but it makes sense that client's implementation would also have one less "if".